apps/builder/app/canvas/features/webstudio-component/webstudio-component.tsx

```diff
@@ -46,9 +49,50 @@ const ContentEditable = ({
   [componentAttribute]: Instance["component"];
 }) => {
   const [editor] = useLexicalComposerContext();
+  const [caretClassName] = useState(() => `a${nanoid()}`);
+  const engineRef = useRef<CssEngine | undefined>();
 
   const ref = useCallback(
     (rootElement: null | HTMLElement) => {
+      if (rootElement === null) {
+        if (engineRef.current !== undefined) {
+          engineRef.current.unmount();
+          engineRef.current = undefined;
+        }
+      }
+
+      if (rootElement !== null) {
+        const elementColor = window.getComputedStyle(rootElement).color;
+
+        const color = colord(elementColor).toRgb();
+        if (color.a < 0.1) {
+          // Apply caret color with animated color
+          const engine = createCssEngine({ name: "text-editor-caret" });
+          engineRef.current = engine;
+
+          // Animation on cursor needed to make it visible on any background
+          engine.addPlaintextRule(`
+
+            @keyframes ${caretClassName}-keyframes {
```

Comment:
This looks more editor specific. Maybe move to text-editor module?

Reply:
Im stupid enough to not found how to get element ref in the text editor ;)

Reply:
You can add class to theme. "root" might work.

Reply:
can't we make this entire css static, rendered once instead of adding it inside every component's render?

Reply:
`document.body.style.caretColor = 'red'`

Reply:
since that works from any root container, it means we can render the css utility class once and add it to some root element when necessary

Reply:
we already have helpersCssEngine in styles.ts which would be a good place for this

Reply:
This is added only if TextEditor is activated so it doesn't affect performance or anything. And basically it's not global but TextEditor only feature. In any 3rd party css engine it will be local css.

Reply:
PS: I moved logic into TextEditor as @TrySound suggested